File: Day-18/Day18.py
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        ans = []
        d = deque()
        for i in range(len(nums)):
            while len(d) > 0 and nums[i] >= nums[d[-1]]:
                d.pop()
            d.append(i)
            if d[0] == i-k:
                d.popleft()
            if i >= k-1:
                ans.append(nums[d[0]])
        return ans

File: Day-18/Day18.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> ans;
        deque<int> d;
        for(int i=0;i<nums.size();i++){
            while(d.size() && nums[d.back()] < nums[i]){
                d.pop_back();   
            }
            d.push_back(i);
            if(d.front() == i-k) d.pop_front();
            if(i >= k-1){
                ans.push_back(nums[d.front()]);
            }
        }
        return ans;
    }
};